tests/realtikvtest/txntest/txn_test.go

```diff
@@ -342,3 +350,185 @@ func TestTxnEntrySizeLimit(t *testing.T) {
+	config.UpdateGlobal(func(conf *config.Config) {
+		conf.TiKVClient.AsyncCommit.SafeWindow = 500 * time.Millisecond
+		conf.TiKVClient.AsyncCommit.AllowedClockDrift = 0
+	})
+
+	// A helper function to determine whether a KV RPC request is handled on TiKV without RPC error or region error.
+	isRequestHandled := func(resp *tikvrpc.Response, err error) bool {
+		if err != nil || resp == nil {
+			return false
+		}
+
+		regionErr, err := resp.GetRegionError()
+		if err != nil || regionErr != nil {
+			return false
+		}
+
+		return true
+	}
+
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	tkPrepare1 := testkit.NewTestKit(t, store)
+	tkPrepare2 := testkit.NewTestKit(t, store)
+	tk1 := testkit.NewTestKit(t, store)
+	tk2 := testkit.NewTestKit(t, store)
+	tkPrepare1.MustExec("use test")
+	tkPrepare2.MustExec("use test")
+	tk1.MustExec("use test")
+	tk2.MustExec("use test")
+
+	tk1.MustExec("create table t (id int primary key, v int)")
+	tk1.MustExec("insert into t values (1, 10), (2, 20)")
+	// Table t2 for revealing the possibility that the issue causing data-index inconsistency.
+	tk1.MustExec("create table t2 (id int primary key, v int unique)")
+	tk1.MustExec("insert into t2 values (1, 10)")
+
+	tkPrepare1.MustExec("set @@tidb_enable_async_commit = 1")
+	tk1.MustExec("set @@tidb_enable_async_commit = 0")
+
+	// Prepare a ts collision (currentTxn.StartTS == lastTxn.CommitTS on the same key).
+	// Loop until we successfully prepare one.
+	var lastCommitTS uint64
+	for {
+		// Reset the value which might have been updated in the previous attempt.
+		tkPrepare1.MustExec("update t set v = 10 where id = 1")
+
+		// Update row 1 in async commit mode
+		assert.NoError(t, failpoint.Enable("tikvclient/beforePrewrite", "pause"))
+		tkPrepapre1Ch := make(chan struct{})
+		go func() {
+			tkPrepare1.MustExec("update t set v = v + 1 where id = 1")
+			tkPrepapre1Ch <- struct{}{}
+		}()
+
+		// tkPrepare2 Updates TiKV's max_ts by reading. Assuming tkPrepare2's reading is just before tk1's BEGIN,
+		// we expect that tk1 have startTS == tkPrepare2.startTS + 1 so that the tk1.startTS == TiKV's min_commit_ts.
+		tkPrepare2.MustQuery("select * from t where id = 1").Check(testkit.Rows("1 10"))
+		tk1.MustExec("begin optimistic")
+
+		assert.NoError(t, failpoint.Disable("tikvclient/beforePrewrite"))
+		select {
+		case <-tkPrepapre1Ch:
+		case <-time.After(time.Second):
+			assert.Fail(t, "tkPrepare1 not resumed after unsetting failpoint")
+		}
+
+		var err error
+		//tk1.MustQuery("select json_extract(@@tidb_last_txn_info, '$.txn_commit_mode')").Check(testkit.Rows("async_commit"))
+		lastCommitTS, err = strconv.ParseUint(tkPrepare1.MustQuery("select json_extract(@@tidb_last_txn_info, '$.commit_ts')").Rows()[0][0].(string), 10, 64)
+		assert.NoError(t, err)
+		currentStartTS, err := strconv.ParseUint(tk1.MustQuery("select @@tidb_current_ts").Rows()[0][0].(string), 10, 64)
+		assert.NoError(t, err)
+		if currentStartTS == lastCommitTS {
```

Comment:
Is it possible to use a failpoint to inject specific ts instead of using a `for` loop?

Reply:
Maybe, but I'm afraid it's hard to claim that our simulation is correct and our simulated transactions can exist in real cases. And IIRC it's not the first time I use this way to simulate ts collision in tests.

Reply:
The case is quite complex, a little bit concern its stability. 
Perhaps it is indeed difficult to encounter in real scenes, which is why it is so difficult to simulate..

Reply:
Yes. Actually I'm kind of doubting about how well the ts-collision cases of async commit/1PC are covered by our random tests... but making these cases right is crucial to the correctness